compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,380 @@
+                });
+            }
+        }
+    }
+    // A conditional exit would have exactly 2 predecessors
+    result.filter(|result| cfg.predecessors(result.block_exit).len() == 2)
+}
+
+/// Computes a cost estimate of a basic block
+/// returns u32::MAX if the block has side-effect instructions
+/// WARNING: the estimates are estimate of the runtime cost of each instructions,
+/// 1 being the cost of the simplest instruction. These numbers can be improved.
+fn block_cost(block: BasicBlockId, dfg: &DataFlowGraph) -> u32 {
+    let mut cost: u32 = 0;
+    for instruction in dfg[block].instructions() {
+        let instruction_cost = match &dfg[*instruction] {
+            Instruction::Binary(binary) => {
+                match binary.operator {
+                    BinaryOp::Add { unchecked }
+                    | BinaryOp::Sub { unchecked }
+                    | BinaryOp::Mul { unchecked } => if unchecked { 3 } else { return u32::MAX },
+                    BinaryOp::Div
+                    | BinaryOp::Mod => return u32::MAX,
+                    BinaryOp::Eq => 1,
+                    BinaryOp::Lt => 5,
+                    BinaryOp::And
+                    | BinaryOp::Or
+                    | BinaryOp::Xor => 1,
+                    BinaryOp::Shl
+                    | BinaryOp::Shr => return u32::MAX,
+                }
+            },
+            // A Cast can be either simplified, or lead to a truncate
+            Instruction::Cast(_, _) => 3,
+            Instruction::Not(_) => 1,
+            Instruction::Truncate { .. } => 7,
+
+            Instruction::Constrain(_,_,_)
+            | Instruction::ConstrainNotEqual(_,_,_)
+            | Instruction::RangeCheck { .. }
+            // Calls with no-predicate set to true could be supported, but
+            // they are likely to be too costly anyways. Simple calls would
+            // have been inlined already.
+            | Instruction::Call { .. }
+            |      Instruction::Load { .. }
+            | Instruction::Store { .. }
+            | Instruction::ArraySet { .. } => return u32::MAX,
+
+            Instruction::ArrayGet { array, index  } => {
+                // A get can fail because of out-of-bound index
+                let mut in_bound = false;
+                // check if index is in bound
+                if let (Some(index), Some(len)) = (dfg.get_numeric_constant(*index), dfg.try_get_array_length(*array)) {
+                    // The index is in-bounds
+                    if index.to_u128() < len as u128 {
+                        in_bound = true;
+                    }
+                }
+                if !in_bound {
+                    return u32::MAX;
+                }
+                1
+            },
+
+            Instruction::Allocate
+            | Instruction::EnableSideEffectsIf { .. }
+            | Instruction::IncrementRc { .. }
+            | Instruction::DecrementRc { .. }
+            | Instruction::MakeArray { .. }
```

Comment:
We want to get a more accurate instruction count here. MakeArray is either several store and add instructions or a runtime loop for large non-nested arrays https://github.com/noir-lang/noir/blob/32f05f4e42ce25ad48d53274be682af104ef0104/compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs#L1714.

Reply:
I put 10 as the cost then, since it seems to be the maximum cost